File: src/Web/Langueedu.Web/Langueedu.Web.Pages/wwwroot/assets/scripts/youtube-player.js
let player = undefined
const namespace = 'Langueedu.Web.Components'

const onPlayerReady = (event) => {
  DotNet.invokeMethodAsync(namespace, 'PlayerReady', event)
}

const onPlayerStateChange = (event) => {
  DotNet.invokeMethodAsync(namespace, 'PlayerStateChange', event)
}

const insertYoutubeApiScript = () => {
  if (document.getElementById('youtubePlayerScript') !== null) {
    return
  }

  var tag = document.createElement('script')
  tag.src = 'https://www.youtube.com/player_api'
  tag.id = 'youtubePlayerScript'
  var firstScriptTag = document.getElementsByTagName('script')[0]
  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag)
}

const init = (videoId) => {
  insertYoutubeApiScript()

  const intervalValue = setInterval(() => {
    if (typeof YT === 'undefined' || typeof (YT || {}).Player === 'undefined') {
      return
    }

    clearInterval(intervalValue)

    player = new YT.Player('LeYoutubePlayer', {
      height: '315',
      width: '100%',
      playerVars: {
        playsinline: 1,
        accelerometer: 1,
        gyroscope: 1,
        rel: 0,
        showinfo: 0,
        // controls: 0,
        'clipboard-write': 1,
        'encrypted-media': 1,
        'picture-in-picture': 1,
      },
      videoId: videoId,
      events: {
        onReady: (events) => onPlayerReady(events),
        onStateChange: (events) => onPlayerStateChange(events),
      },
    })
  }, 100)
}

const currentTime = () => {
  if (!player) {
    return 0
  }

  return player.playerInfo.currentTime
}

const playVideo = () => {
  if (player) {
    player.playVideo()
  }
}

const pauseVideo = () => {
  if (player) {
    player.pauseVideo()
  }
}

const stopVideo = () => {
  if (player) {
    player.stopVideo()
  }
}

const loadVideoById = (videoId, startSeconds = 0) => {
  if (player && videoId) {
    player.loadVideoById(videoId, startSeconds)
  }
}

const scrollIntoView = (elementId) => {
  if (elementId) {
    document.getElementById(elementId).scrollIntoView({
      behavior: 'auto',
      block: 'center',
      inline: 'center',
    })
  }
}

window.youtubePlayer = {
  init,
  currentTime,
  playVideo,
  pauseVideo,
  stopVideo,
  loadVideoById,
  scrollIntoView,
}

File: src/Web/Langueedu.Web/Langueedu.Web.Components/Services/YoutubePlayer.cs
using System.Windows.Input;
using Langueedu.Web.Components.Interfaces;
using Microsoft.JSInterop;

namespace Langueedu.Web.Components.Services;

public class YoutubePlayer : IYoutubePlayer
{
  private readonly IJSRuntime _jSRuntime;

  public YoutubePlayer(IJSRuntime jSRuntime)
  {
    _jSRuntime = jSRuntime;
  }

  public static ICommand OnPlayerReady;
  public static ICommand OnPlayerStateChange;

  public async Task InitYoutube(string videoId)
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.init", videoId);

  }

  public async Task<double> CurrentTime()
  {
    return await _jSRuntime.InvokeAsync<double>("youtubePlayer.currentTime");
  }
  public async Task ScrollIntoView(string elementId)
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.scrollIntoView", elementId);
  }

  public async Task PlayVideo()
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.playVideo");
  }

  public async Task PauseVideo()
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.pauseVideo");
  }

  public async Task StopVideo()
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.stopVideo");
  }

  public async Task LoadVideoById(string videoId, short startSeconds = 0)
  {
    await _jSRuntime.InvokeVoidAsync("youtubePlayer.loadVideoById", videoId, startSeconds);
  }

  [JSInvokable]
  public static void PlayerReady(object data)
  {
    OnPlayerReady?.Execute(null);
  }

  [JSInvokable]
  public static void PlayerStateChange(object data)
  {
    OnPlayerStateChange?.Execute(null);
  }
}
